Allow IP detection via qemu-agent (proxmox salt-cloud)

MasonEdgel opened the pull request:
### What does this PR do?
Allows salt-cloud to create QEMU VMs and get their IP via the agent when using a Proxmox cloud.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/47771

### Previous Behavior
Would silently error out during a clone, as no IP address was determined for the new VM

### New Behavior
Checks if new VM is of type `qemu`, and uses the qemu agent to determine the IP.  This requires that qemu-guest-agent be installed to the source VM.

### Tests written?
No

### Commits signed with GPG?
No

This isn't really a solution so much as a workaround, looks like the develop branch may take an entirely different approach to the cloning process and I haven't had time to see if the issue persists in that branch.  These are simply the changes I made to my local installation in order to get it working for the time being.  

There is likely a more elegant way to detect whether the qemu-agent is running on the target VM than sleeping for 20 seconds, but I can't think of a way other than checking the URL repeatedly, which still won't work if qemu-agent isn't installed.

EDIT:  After a bit more review, looks like the develop branch doesn't actually change this behavior, so this may still be relevant there.

MasonEdgel committed:
Allow IP detection via qemu-agent

MasonEdgel committed:
Fix typo

gtmanfred commented on salt/cloud/clouds/proxmox.py:
```diff
@@ -585,6 +586,23 @@ def create(vm_):
         return {'Error': 'Unable to start {0}, command timed out'.format(name)}
 
+    # For QEMU VMs, we can get the IP Address from qemu-agent
+    if nodeType == 'qemu':
```
This should behind a configuration gate, we don't require people to have qemu-guest-agent installed now, so we shouldn't force them to do this.

Also, this should probably go against develop instead of 2018.3, as it would be a new feature.

gtmanfred commented on salt/cloud/clouds/proxmox.py:
```diff
@@ -585,6 +586,23 @@ def create(vm_):
+    if nodeType == 'qemu':
+        # We have to wait for a bit for qemu-agent to start
+        log.debug("Waiting for qemu-agent to start...")
+        time.sleep(20)
```
I don't like this, we should just use the `cloud.wait_for_fun` utils module to know when the node is available and ready to return.

https://github.com/saltstack/salt/blob/2018.3/salt/cloud/clouds/openstack.py#L713  Here is an example.

MasonEdgel committed:
Move to config, use wait_for_fun

MasonEdgel commented:
@gtmanfred, I've made the suggested changes.  I didn't want to hijack the `agent` config setting, in case someone wanted to enable it but still use static IP, so I added an `agent_get_ip` setting..  I've also moved from `time.sleep` to `wait_for_fun`.

rallytime commented:
Hi @MasonEdgel - There are some merge conflicts with this change. Can you rebase this against the current HEAD of develop and fix up the conflicts?

MasonEdgel committed:
Merge branch 'develop' into cloud-proxmox-qemu-ip-fix

MasonEdgel commented:
Sorry all! I'm still getting the hang of this PR thing...  I _think_ I've fixed the history issues.  This PR should now only contain my current changes on top of the develop branch.

gtmanfred commented:
Awesome, we are almost there, this is looking really good now, we just need to add `agent_get_ip` to the getting started with proxmox doc and then you have a couple pylint tests.

https://jenkinsci.saltstack.com/job/pr-lint/job/PR-50189/9/warnings52Result/new/

https://github.com/saltstack/salt/blob/develop/doc/topics/cloud/proxmox.rst

Thanks!

MasonEdgel committed:
Jenkins lint cleanup

MasonEdgel committed:
Add reference to doc

MasonEdgel committed:
Remove missed whitespace

gtmanfred commented:
Thanks @MasonEdgel this looks great.

MasonEdgel committed:
Code Climate fixes

terminalmage commented:
Ported to master in https://github.com/saltstack/salt/pull/56710
